### PR TITLE
Add snapshots

### PR DESCRIPTION
**Pull request by brohamgoham**
## Description
Added errors for snapshot creation, and beginning the ABCI addition of listing snapshots to bootstrap new nodes

## Related Issue

Fixes # (issue)

## Breaking Change

- [x] Yes
- [ ] No

## Checklist:

- [x] I have read and followed the CONTRIBUTING guidelines for this project.
- [x] I have added or updated tests and they pass.
- [ ] I have added or updated documentation and it is accurate.
- [x] I have noted any breaking changes in this module or downstream modules.

### CODE REVIEW

File: src/many/src/message/error.rs [hunk 1]
```diff
@@ -105,7 +105,14 @@ many_error! {
        -9: InvalidAttribtueId as invalid_attribute_id(id) => "Unexpected attribute ID: {id}.",
       -10: InvalidAttributeArguments as invalid_attribute_arguments()
             => "Attribute does not have the right arguments.",
-      -11: AttributeNotFound as attribute_not_found(id) => "Expected attribute {id} not found.",
+      -11: AttributeNotFound as attribute_not_found(id) 
+            => "Expected attribute {id} not found.",
+      -12: SnapshotNotFound as snapshot_not_found(id)
```
Comment (hansl) [inline src/many/src/message/error.rs:110]: These should be in `_1000_abci_backend` or frontend.

File: src/many/src/message/error.rs [hunk 2]
```diff
@@ -153,6 +154,12 @@ many_error! {
             => "The message's timestamp is out of the accepted range of the server.",
     -1006: RequiredFieldMissing as required_field_missing(field)
             => "Field is required but missing: '{field}'.",
+    -1007: SnapshotNotFound as snapshot_not_found(id)
```
Comment (hansl) [inline src/many/src/message/error.rs:157]: These are not generic errors, they're part of the 1000 attribute and should be defined there as attribute errors.

File: src/many/src/server/module/_1000_abci_backend.rs [hunk 3]
```diff
@@ -51,6 +52,67 @@ pub struct AbciCommitInfo {
...
+#[cbor(map)]
+pub struct Snapshot {
+    #[n(0)]
+    pub path: PathBuf,
+    #[n(1)]
+    pub height: u64,
+    #[n(2)]
+    pub hash: Vec<u8>,
+ //   #[n(3)]
+ //   pub chunks: u64,
```
Comment (hansl) [inline src/many/src/server/module/_1000_abci_backend.rs:65]: ?

**Review by hansl — CHANGES_REQUESTED** [review]

File: src/many/src/server/module/_1000_abci_backend.rs (hunk 3 above)
Comment (brohamgoham) [reply src/many/src/server/module/_1000_abci_backend.rs]: Added chunks back to reflect TM Snapshot struct

File: src/many/src/message/error.rs (hunk 2 above)
Comment (hansl) [reply src/many/src/message/error.rs:157]: Please address this. `-1000` is not the attribute ID. Take a look here to see how to define attribute specific errors: https://github.com/the-lifted-initiative/many-rs/blob/main/src/many/src/server/module/_9_account/errors.rs#L3-L9

**Review by hansl — CHANGES_REQUESTED** [review]
One change but otherwise LGTM.

File: src/many/src/server/module/_1000_abci_backend.rs [hunk 4]
```diff
@@ -51,6 +52,86 @@ pub struct AbciCommitInfo {
...
+    pub fn new(snapshots: Vec<Snapshots>) -> Self {
+        AbciListSnapshot {
+            snapshots,
+        }
+    }  
+}
+
+#[derive(Encode, Decode)]
+#[cbor(map)]
+pub struct AbciOfferSnapshot {
```
Comment (hansl) [inline src/many/src/server/module/_1000_abci_backend.rs:113]: Question; do we want to show the height here?

**Comment by Denommus** [comment]
PR has conflicts

**Pull request closed without merging**